ci-operator/step-registry/observers/ocp-baremetal-qe/observers-ocp-baremetal-qe-commands.sh

```diff
@@ -0,0 +1,411 @@
+            <success message="">Host #${HOST} ($HOSTNAME) should be reachable and respond on the SSH port</success>
+          </testcase>
+          <testcase name="install should succeed: host booted expected live image">
+            <success message="">Host #${HOST} ($HOSTNAME) should boot the expected live image</success>
+          </testcase>
+          <testcase name="install should succeed: host installed CoreOS">
+            <failure message="">Host #${HOST} ($HOSTNAME) should boot the installed OS from Disk</failure>
+          </testcase>
+          <testcase name="install should succeed: overall">
+            <failure message="">openshift cluster install failed overall</failure>
+          </testcase>
+        </testsuite>
+EOF
+    else
+      cat >>"${ARTIFACT_DIR}/junit_install_${HOST}.xml" <<EOF
+      <testsuite name="cluster install" tests="3" failures="0">
+          <testcase name="install should succeed: host reachable">
+            <success message="">Host #${HOST} ($HOSTNAME) should be reachable and respond on the SSH port</success>
+          </testcase>
+          <testcase name="install should succeed: host booted expected live image">
+            <success message="">Host #${HOST} ($HOSTNAME) should boot the expected live image</success>
+          </testcase>
+          <testcase name="install should succeed: host installed CoreOS">
+            <success message="">Host #${HOST} ($HOSTNAME) should boot the installed OS from Disk</success>
+          </testcase>
+          <testcase name="install should succeed: overall">
+            <success message="">openshift cluster install succeeded</success>
+          </testcase>
+        </testsuite>
+EOF
+    fi
+    cat >>"${ARTIFACT_DIR}/junit_install_${HOST}.xml" <<EOF
+  </testsuite>
+EOF
+  done < "$input"
+  echo "JUnit reports created, exiting"
+  fi
+  exit 0
+}
+
+function isNodeReachable(){
+  # Check if node is reachable poking SSH port
+  local host="${1}"
+  ssh_port=$((12000 + $host))
+  status=$(timeout 5s nc ${AUX_HOST} "${ssh_port}" || true;)
+  if [[ $status == *"SSH"* ]]; then
+      echo $NODE_IS_REACHABLE
+  else
+      echo $NODE_IS_UNREACHABLE
+  fi
+}
+
+# 3. For each node, wait up to 7 minutes until something like curl -x $PROXY $HOST_IP:22 returns the version of OpenSSH.
+
+function isNodeAlive(){
+  local bmhost="${1}"
+  # shellcheck disable=SC1090
+  . <(echo "$bmhost" | yq e 'to_entries | .[] | (.key + "=\"" + .value + "\"")')
+  echo "Starting isNodeAlive for ${host}"
+  for i in $(seq 1 $MAX_RETRY); do
+    printf "%s: Checking SSH connectivity for %s %s/${MAX_RETRY}\n" "$(date --utc --iso=s)" "${ip}" "${i}"
+    status="$(isNodeReachable "$host" || true;)"
+    if [[ $status == $NODE_IS_REACHABLE ]]; then
+      echo "Node ${host} alive, waiting for services to come up..."
+      # journalRecord may not work even if SSH check passed
+      sleep 60
+      journalRecord $bmhost &
+      checkBootedImage "boot" $bmhost &
```

Comment:
do you ever wait for all these subprocesses to join?

Reply:
no, what's the purpose in doing so ?

Reply:
after you add the `wait` everywhere needed, any 'deadlocks'-like issues should become clearer (for example, the one you mentioned for the missing catch of the last reboot).

Try to have something like:

Main Process => One sub-process per Host => An infinite loop with a switch/case that call different functions based on the value of a $STATE variable.

The $STATE variable will track the evolution of an FSM that represents the lifecycle of the node.

If you still need to fork within a host's sub-process, you can do it in that switch/case and add a wait specific to the PID of the processes you need to join.

Reply:
Also add better in-line comments to let future devs able to debug this script and follow the flow of the observers.

For me, it works in general, just need cleanup and handling the sub-processes lifecycle correctly

Reply:
is the `wait` in the `cleanup` function enough?

Reply:
`wait` should be called everytime you instantiate N processes and want to wait for them to join. Also look at the parameters you can pass to wait on the version of bash we use: you might need to catch the PIDs of the subprocesses and then wait/join only a specific subset of them.